### PR TITLE
✨ Add internal/db/storage/blob

### PR DESCRIPTION
**Pull request by rinx**
Signed-off-by: Rintaro Okamura <rintaro.okamura@gmail.com>

### Description:

- Added internal/db/storage/blob & internal/db/storage/blob/s3 packages
- already tested using blob_test.go

### Related Issue:

Nothing.

### How Has This Been Tested?:

Running blob_test.go

### Environment:

- Golang Version: 1.14.3
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.11.5

### Types of changes:

- [ ] Bug fix [type/bug]
- [X] New feature [type/feature]
- [ ] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [X] Have you written new tests for your core changes, as applicable?
- [X] Have you successfully ran tests with your changes locally?

### Checklist:

- [X] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [X] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [X] I have added tests and benchmarks to cover my changes.
- [X] I have ensured all new and existing tests passed.
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: internal/config/blob.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	switch bst {
+	case S3:
+		return "s3"
+	}
+	return "unknown"
+}
+
+func AtoBST(bst string) BlobStorageType {
+	switch strings.ToLower(bst) {
+	case "s3":
```
Comment (kpango) [inline internal/config/blob.go:38]:
```suggestion
	case S3.String():
```

File: internal/config/blob.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+// Package config providers configuration type and load configuration logic
+package config
+
+import "strings"
+
+type BlobStorageType uint8
+
+const (
+	S3 BlobStorageType = iota
```
Comment (kpango) [inline internal/config/blob.go:25]: It's better to start iota from 1
```suggestion
	S3 BlobStorageType = 1+iota
```

File: internal/db/storage/blob/s3/s3.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	s := new(s3session)
+	for _, opt := range append(defaultOpts, opts...) {
+		opt(s)
+	}
+
+	return s
+}
+
+func (s *s3session) URLOpener() *s3URLOpener {
+	session := session.New(
```
[the remark this answers is not in the document]
Comment (rinx) [reply internal/db/storage/blob/s3/s3.go:50]: nice catch! :dog2:

File: internal/db/storage/blob/s3/s3.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+package s3
+
+import (
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"gocloud.dev/blob/s3blob"
+)
+
+type s3URLOpener = s3blob.URLOpener
```
Comment (kpango) [inline internal/db/storage/blob/s3/s3.go:26]: this file is already s3 package if we use this type we have to write `s3.s3URLOpener` it looks not good.
```suggestion
type URLOpener = s3blob.URLOpener
```

File: internal/db/storage/blob/s3/s3.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+type s3session struct {
```
Comment (kpango) [inline internal/db/storage/blob/s3/s3.go:28]: same
```suggestion
type session struct {
```

File: internal/db/storage/blob/s3/s3.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+	endpoint        string
+	region          string
+	accessKey       string
+	secretAccessKey string
+	token           string
+}
+
+type S3Session interface {
```
Comment (kpango) [inline internal/db/storage/blob/s3/s3.go:36]: same
```suggestion
type Session interface {
```

File: internal/db/storage/blob/s3/s3.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+	URLOpener() *s3URLOpener
+}
+
+func NewS3Session(opts ...Option) S3Session {
```
Comment (kpango) [inline internal/db/storage/blob/s3/s3.go:40]: same
```suggestion
func NewSession(opts ...Option) Session {
```

**Comment by rinx** [comment]
@kpango Thanks for your suggestions. I've revised them.

**Comment by kpango** [comment]
LGTM

**Pull request merged**